File: unid-node-wallet-sdk/src/core/document.ts
import { DIDDocument, ServiceEndpoint, DidPublicKey } from "./interfaces/did-document";

/**
 */
export class UNiDDidDocument {
    /**
     */
    private _document: DIDDocument

    /**
     * @param document 
     */
    constructor(document: DIDDocument) {
        this._document = document
    }

    /**
     */
    public get document(): DIDDocument {
        return this._document
    }

    /**
     */
    public get identifier(): string {
        return this._document.id
    }

    /**
     */
    public get publicKeys(): Array<DidPublicKey> {
        return this._document.publicKey
    }

    /**
     */
    public get services(): Array<ServiceEndpoint> {
        return this._document.service
    }

    /**
     * @param keyId 
     * @returns
     */
    public getPublicKey(keyId: string): DidPublicKey {
        const ks = Object.assign([], this.publicKeys) as Array<DidPublicKey>
        const k  = ks.filter((k) => { return k.id === `#${ keyId }` }).pop()

        if (! k) {
            throw new Error()
        }

        return k
    }
}

File: src/unid/did/document.rs
use super::interfaces::did_document::*;

#[derive(Debug)]
pub struct UNiDDidDocument {
    pub document: DidDocument,
}

impl UNiDDidDocument {
    pub fn new(document: DidDocument) -> Self {
        UNiDDidDocument { document }
    }
    pub fn document(&self) -> DidDocument {
        self.document.clone()
    }
    pub fn identifier(&self) -> String {
        self.document.id.to_string()
    }
    pub fn public_keys(&self) -> Vec<DidPublicKey> {
        self.document.public_key.clone()
    }
    pub fn services(&self) -> Vec<::serde_json::Value> {
        self.document.service.clone()
    }

    pub fn get_public_key(&self, key_id: String) -> DidPublicKey {
        let ks: Vec<DidPublicKey> = self.document().public_key;
        let filtered_ks = ks
            .iter()
            .cloned()
            .filter(|x| x.id == format!("#{}", key_id))
            .collect::<Vec<DidPublicKey>>();
        let k = filtered_ks.first();
        assert!(k.is_some());
        k.unwrap().clone()
    }
}